CIPs/0016.md

```diff
@@ -0,0 +1,89 @@
+# CIP [0016]: Universal celo protocol URI scheme for QRs/deep links
+
+- Date: 2020-07-27
+- Author: @yorhodes
+- Status: WIP
+
+## Overview
+
+Create a standard URI scheme for encoding interacting with the Celo protocol for various consumers including SDKs, mobile/web wallets, and end users.
+Facilitate development of QR systems by standardizing usage and unifying scheme for interoperability.
+
+## Goals
+
+- establish URI standard compatible with mobile app intents that optimizes for **memory** and **performance**
+- enable simple construction of URIs by humans without deep Celo protocol understanding
+- enable simple construction of QRs by code from arbitrary Celo transactions
+
+## Proposed Solution
+
+### Syntax
+Celo protocol URLs contain "celo" in their schema (protocol) part and are constructed as follows:
+
+    request                 = "celo:" [ prefix "-" ] target [ "/" endpoint ] [ "?" parameters ]
+    prefix                  = "pay" / "request" / "call" / "send"
+    
+    target                  = E164_PHONE_NUMBER / registry_id / celo_address
+    registry_id             = "StableToken" / "GoldToken" / "LockedGold" / ...
+    celo_address            = ( "0x" 40*HEXDIG )
+    
+    endpoint                = token / function
+    token                   = "cUSD" / "CELO" / ...
+    function                = function_name "(" [ function_args ] ")"
+    function_name           = STRING
+    function_args           = function_arg *( "," function_arg )
+    function_arg            = ( NUMBER / STRING ) [ ":" TYPE ]
+
+    parameters              = parameter *( "&" parameter )
+    parameter               = key "=" value
+    key                     = "value" / "gas" / "gasPrice" / "feeCurrency" / "gatewayFee" / "gatewayFeeRecipient" / "localAmount" / "localCurrencyCode" / ...
+    value                   = NUMBER / STRING / celo_address
+
+### Semantics
+
+`TYPE` is a standard ABI type name, as defined in [Ethereum Contract ABI specification](https://solidity.readthedocs.io/en/develop/abi-spec.html). `STRING` is a URL-encoded unicode string of arbitrary length, where delimiters and the percentage symbol (`%`) are mandatorily hex-encoded with a `%` prefix.
+
+`E164_PHONE_NUMBER` is a standard E164 universal phone number, as defined by the International Telecommunication Union [E164 Number Specification](https://www.itu.int/rec/T-REC-E.164).
+
+`prefix` is used for differentiating between user and contract `target`s and actions. `pay` and `request` are for user targets and intended for use with phone numbers or addresses and the `token` endpoint. `call` & `send` are for contract targets and intended for use with the registry/addresses and the `function` endpoint. 
```

Comment:
So just so I understand this - `call` and `send` represent view calls vs sending a transaction on a contract? Also, I think there's a little bit of weirdness around `pay` and `request` - pay directly maps to sending a transaction whereas request is an application level construct that is stored off chain. I wonder if we should differentiate that somehow

Reply:
yeah it's apparent that the prefix system doesn't make much sense as it's mixing a few different categories
I figured these would be nice to haves for deep-linking to specific application state